Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,217 @@
+When an intrinsic namespace is introduced code will likely be implemented on different platforms at different times.
+
+It is explicitly OK for ARM64 to implement a namespace which may be implemented on `ARM32` w/o `ARM32` implementation.
+The opposite is also true.
+
+Any platform which does not implement a namespace, must:
+
++ Return `false` for `IsSupported`
++ `throw` on any unimplemented function
+
+#### Partial implementation of intrinsic namespace
+
++ It is preferred that `IsSupported` represents the state of an entire namespace
++ It is certainly required at time of a release, `IsSupported` represents the state of an entire namespace
++ Allow for writing tests and implementing code, `IsSupported == false` may be used when namespace is partially implemented
+
+#### Addition of new intrinsics to existing namespace
+When an intrinsic namespace is introduced, a best guess will be made about the set of useful intrinsics.
+However, the intrinsics will be added as needed or requested for various optimization efforts.
+
++ The set of intrinsics instructions supported by a namespace should only be allowed
+if all platforms reporting `IsSupported` are supported (at least at time of release)
++ It may be preferable to add new namespaces in cases where this creates significant burden
+
+### Half precision floating point
+
+This document will refer to half precision floating point as `Half`.
+
++ Machine learning and Artificial intelligence often use `Half` type to simplify storage and improve processing time.
++ CoreCLR and `CIL` in general do not have general support for a `Half` type
++ There is an open request to expose `Half` intrinsics
++ ARM64 will introduce `System.Runtime.Intrinsics.Half` to support this request 
++ `Half` will be defined as:
+
+```
+    [StructLayout(LayoutKind.Sequential, Size = 2)]
+    public struct Half : struct {}
+``` 
+
+#### ARMv8 Half precision support
+
+ARMv8 baseline support for `Half` is limited.  The following operations are supported
+
++ Loads and Stores
++ Conversion to/from `Float`
++ Widening from `Vector128<Half>` to two `Vector128<Float>`
++ Narrowing from two `Vector128<Float>` to `Vector128<Float>`
+
+It is presumed that even this minimal support could be helpful
+
+Optional `Half` extensions add more complete support
+
+#### `Half` and ARM64 ABI
+
+The proposed `Half` implementation will treat `Half` as raw bits.  
+It is likely therefore that this will not conform to the ARM64 ABI specifically with respect to the HFA
+calling convention.  This will hinder use of `Half` in `PInvoke` calls.  Resolving this ABI issue will
+be a low priority.
+
+### Scalable Vector Extension Support
+
+`SVE`, the Scalable Vector Extension introduces its own complexity.  
+
+The extension 
++ Creates a new set of SVE registers 
++ Each register has a platform specific length
+  + Any multiple of 128 bits
+
+Therefore implementation will not be trivial.
+
++ Register allocator may need changes
```

Comment:
Arm claims in documentation that implementation of register sizes should be transparent to applications what is totally unclear to me.
Perhaps we should reach to Arm and ask what they mean that register sizes should be transparent to software using SVE

Reply:
By the time it is feasible to implement and accurately test this, I am hoping for clearer docs.  We will certainly have a better understanding by that time.

Reply:
I will reword this section to make sure it is clear the SVE design is incomplete.

Implementation will depend on having a Architecturally accurate model and/or functional Architecturally compliant hardware for testing.  AFAIK, SVE spec is too new to expect this soon.